### PR TITLE
Make platform independent

### PR DESCRIPTION
**Pull request by daniel-mohr**
At least most of #31 should be fixed with this pull request. For the given test cases the solution should be fine. Some problems addressed in #31 are not solved.

### CODE REVIEW

**Comment by sadielbartholomew** [comment]
@daniel-mohr FYI I have just pushed some commits I made yesterday and forgot to push upstream, one of which makes a change so that the linting job should now pass. I will review this tomorrow.

**Comment by daniel-mohr** [comment]
@sadielbartholomew I updated/merged your new master into my branch make_platform_independent. Unfortunately the tests do not run again and I also do not find a way to initiate a re-run. Any ideas? But the merge was simple.

**Comment by sadielbartholomew** [comment]
Thanks @daniel-mohr.

>  Unfortunately the tests do not run again and I also do not find a way to initiate a re-run. Any ideas?

With the current settings for the workflows, all (I think) having set:

https://github.com/NCAS-CMS/cfunits/blob/eca4441db8a124f4fac40b5011b359e40871c940/.github/workflows/run-test-suite.yml#L9-L11

the easiest way, though not particularly elegant, to re-trigger the jobs is to open and close the PR to hit the `reopened` category, which is what I have done for a while (feel free to try it and see). If you know a more elegant but equally simple way to re-trigger the jobs then please do let us know and/or update that part of the workflows to effect the necessary change in Actions behaviour.

**Review by sadielbartholomew — COMMENTED** [review]
Generally this looks great:

* though I don't have a Mac accessible to test on directly, from the CI jobs I can see that, as far as the test suite can detect, cfunits will work for all of the platforms after this PR;
* locally on my Linux and Windows machines the test suite also passes;
* the solution seems sensible (because I don't like to trust any test suite to pick up on any possible issue) after I cross-referenced with an inspection of `np.sctypeDict` for all of the various `dtype` and `dtype.char` possibilities.

Before I approve and merge the PR:

1. I have made a few minor comments regarding the code style, please see those.
2. Now that you have made cfunits platform-independent, we can update the `setup.py` to indicate this. If you like (and if not, that's not a problem since I can do it after this PR is merged) please go ahead and update the `platforms` and `classifiers` -> `Operating System` items in the `setup` function there as a new commit (see the [equivalent values for the `cfdm` library](https://github.com/NCAS-CMS/cfdm/blob/24e963428d2177fae47a459aef58942e5b9a6b8f/setup.py#L150-L167) which has a similar `setup.py` where platform independence is already indicated, as a guide).

3. I had one general question, namely it looks like there is similar conditional logic based on the data type properties of `x` in the subsequent (`else`) case where the operation is not `inplace`, i.e:

https://github.com/NCAS-CMS/cfunits/blob/eca4441db8a124f4fac40b5011b359e40871c940/cfunits/units.py#L2225-L2230

Is there a reason why that part doesn't need to be adapted in a similar way to the corresponding `if` block for platform-independence too? Note that it may not be covered by the test suite (yet), but we should still make sure it will also work in a platform-independent way (and eventually get it included in the test suite, naturally).

File: cfunits/units.py
Comment (sadielbartholomew) [inline cfunits/units.py:2219]: Sorry I forgot to submit one comment I meant to attach to the previous review: instead of checking for presence in a one-item list, it is probably simpler and clearer to use:

```suggestion
                    if x.dtype == numpy.int32:
```

unless there is some reason you know that it isn't a good idea to do that? (The same applies to the same form of statement used in the following lines.)

File: cfunits/units.py
Comment (daniel-mohr) [reply cfunits/units.py:5]: OK. But I do not believe in the performance benefit. A few years ago I analyzed it quickly. It is much more important to structure a module in adequate submodules and not importing all submodules. Maybe on some implementations (e. g. cython) it makes a difference.

But anyway, due to my comments to the issue #31 I provide a better code without importing numpy.

File: cfunits/units.py
Comment (daniel-mohr) [reply cfunits/units.py:2219]: I was thinking that we should do the same for many different dtypes. Therefore are started using a list.

But anyway, due to my comments to the issue #31 I provide a better code without importing numpy.

**Comment by daniel-mohr** [comment]
@sadielbartholomew I commented your 'code comments', but after a new/better commit I cannot find them. But anyway, due to my better solution (based on my comments in #31) they are obsolete.

I will react on your other comments later (tomorrow or ...).

File: cfunits/units.py
Comment (sadielbartholomew) [reply cfunits/units.py]: Fair enough, though @davidhassell may have other reasons for choosing that convention that is used across related libraries as well (cfdm, cf-python) that were more important than performance. He might be able to elaborate when he returns form leave...

File: cfunits/units.py
Comment (sadielbartholomew) [reply cfunits/units.py]: That makes sense. No problem.

**Comment by sadielbartholomew** [comment]
Thanks @daniel-mohr, that sounds good. Take your time, there is unlikely to be any major or indeed minor changes going into `cfunits` across the next week, as far as I am aware, so there is little chance for significant merge conflicts etc.

**Comment by daniel-mohr** [comment]
Regarding the code: 
https://github.com/NCAS-CMS/cfunits/blob/eca4441db8a124f4fac40b5011b359e40871c940/cfunits/units.py#L2225-L2230

[I'm also not glad about it.](https://github.com/NCAS-CMS/cfunits/issues/31#issuecomment-882750822) But I do not want to decide this. Let's wait for @davidhassell .

**Comment by daniel-mohr** [comment]
Adapting setup.py  regarding windows: Should we only add "Operating System :: Microsoft :: Windows" or should we do "Operating System :: OS Independent"? (I think it is as independent as udunits is available -- it's python!)

**Comment by sadielbartholomew** [comment]
> Adapting setup.py regarding windows: Should we only add "Operating System :: Microsoft :: Windows" or should we do "Operating System :: OS Independent"? (I think it is as independent as udunits is available -- it's python!)

`"Operating System :: OS Independent"` please, it is more concise than listing out all of the possibilities (Linux, MacOS and now Windows)!

**Comment by sadielbartholomew** [comment]
@daniel-mohr, one small thing to mention, FYI: I am quickly putting in a PR to do some automatic sorting of import statements across some libraries, including `cfunits` (#33), so there is the potential for a small merge conflict from that, though I see you aren't applying any imports at this point, so hopefully there won't be any at all.

**Comment by daniel-mohr** [comment]
'setup.py' is now adapted.

I have no intention to add further code to this PR. So, from my point of view a merge of this PR could be done.

**Comment by sadielbartholomew** [comment]
Thanks @daniel-mohr. The new commits look good to me, with the `setup.py` being adapted correctly etc.

Since @davidhassell is due back from leave tomorrow, and it would be good to have a second opinion on the changes here because the code touched in the `units` module is quite critical to `cfunits` working properly, I will do a final review later today (since you have changed approach) but wait to merge until David has signed this off too, likely in the next few days or so.

**Comment by davidhassell** [comment]
Hi @daniel-mohr and @sadielbartholomew  - lot's has gone since I've been away  - great. I'll have a look at the dtype issues today.

**Review by sadielbartholomew — COMMENTED** [review]
Hi @daniel-mohr, thanks again for all of your efforts towards this PR. @davidhassell and I looked through and discussed it today and overall we are happy with it, but we want to do some final checks just to ensure it is watertight and we haven't quite finished those, so please bear with us and we will get back to you next week (note we have university closure days on Monday and Tuesday so it won't be until mid-week that we can finish our review).

**Review by sadielbartholomew — APPROVED** [review]
Thanks for waiting for a final review, @daniel-mohr. @davidhassell and I completed our checks and have agreed as far as we can determine that all the changes are sound. (Some of the code is quite clever and we were working out why certain changes were made and the intended behaviour with inplace operations. Just trying to catch up with your thinking, really!) We are very happy to merge this now.

Thanks for all of your fantastic work towards this PR and the corresponding issue! We are going to do a release imminently (probably next week) so that there is a new version inclusive of this update, which should therefore work on Windows. Please keep an eye out for that. Also, any further contributions to cfunits or the other NCAS-CMS libraries would be very welcome indeed, if you have the bandwidth!

**Comment by davidhassell** [comment]
Version 3.3.4 now available in PyPi. Thanks @daniel-mohr and @sadielbartholomew for your work on this. Much appreciated.

**Pull request merged**
